# Custom Vision

In [1]:
!pip install azure-cognitiveservices-vision-customvision  # 마이크로소프트가 만든 Custom Vision이 있다.

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 62 kB 647 kB/s 
     |████████████████████████████████| 85 kB 4.6 MB/s 
     |████████████████████████████████| 178 kB 56.7 MB/s 
     |████████████████████████████████| 41 kB 675 kB/s 


In [2]:
from azure.cognitiveservices.vision.customvision.training import CustomVisionTrainingClient
from azure.cognitiveservices.vision.customvision.prediction import CustomVisionPredictionClient
from azure.cognitiveservices.vision.customvision.training.models import ImageFileCreateBatch, ImageFileCreateEntry, Region
from msrest.authentication import ApiKeyCredentials
import os, time, uuid

In [3]:
# 아래와 같은 값을 찾아서 넣어준다.
ENDPOINT_Training = 'https://labuser2custom.cognitiveservices.azure.com/'  # Training과 Prediction을 따로 따로 만든다.
ENDPOINT_Prediction = 'https://labuser2custom-prediction.cognitiveservices.azure.com/'

training_key = '103e1f4cfae64d838de6042b90877542'
prediction_key = '88bae04998824c67bf1700a6183622ad'
prediction_resource_id = '/subscriptions/7ae06d59-97e1-4a36-bbfe-efb081b9b03b/resourceGroups/RG2/providers/Microsoft.CognitiveServices/accounts/labuser2custom'

In [4]:
credentials = ApiKeyCredentials(in_headers={"Training-key": training_key})  # 보안때문에 변경이 된것 같다.
trainer = CustomVisionTrainingClient(ENDPOINT_Training, credentials)

In [5]:
print("Creating project...")
project = trainer.create_project("Labuser2 Project")

Creating project...


In [6]:
Jajangmyeon_tag = trainer.create_tag(project.id, "Jajangmyeon")
Champon_tag = trainer.create_tag(project.id, "Champon")
Tangsuyug_tag = trainer.create_tag(project.id, "Tangsuyug")

In [7]:
import time  # 시간 객체를 불러온다.

In [8]:
print('Training....')
iteration = trainer.train_project(project.id)

while(iteration.status != 'Completed'):
  iteration = trainer.get_iteration(project.id, iteration.id)
  print('Training status' + iteration.status)

  time.sleep(5) # 5초에 한번씩 체크를 해준다.
print('Done!')

Training....
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Training statusTraining
Tra

# End(2022년 10월 17일)

In [9]:
from azure.cognitiveservices.vision.customvision.prediction import CustomVisionPredictionClient

In [10]:
prediction_credentials = ApiKeyCredentials(in_headers={"Prediction-key": prediction_key})
predictor = CustomVisionPredictionClient(ENDPOINT_Prediction, prediction_credentials)  # 예측이 가능한 개체다라고 생각하면 된다.
# predictor = CustomVisionPredictionClient(prediction_key, endpoint = ENDPOINT)

In [11]:
target_image_url = 'https://www.koreanbapsang.com/wp-content/uploads/2011/05/DSC1677-1-e1646885376296.jpg'
result = predictor.classify_image_url(project.id, 'greatwall', target_image_url)  # 예측하는 객체를 이미지의 url을 가지고 분석을 하겠다는 것이다. 그리고 배포하는 것의 이름이 greatwall이다.

In [20]:
for prediction in result.predictions:
  print('\t' + prediction.tag_name + " : {0:0.2f}%".format(prediction.probability * 100)) # 소수점 둘째자리 까지 나타내는 것이다.

# 개별적으로 평가를 하였기 때문에 3개의 전체 합이 100%가 안된다.(Jajangmyeon, Champon, Tangsuyug)

	Jajangmyeon : 98.40%
	Champon : 0.64%
	Tangsuyug : 0.24%


* target_image_url에 있는 짜장면이 짜장면인지 테스트를 해보았을 때, Jajangmyeon의 probability는 98.40%, Champo의 probabiliy는 0.64%, Tangsuyug의 probability는 0.24%가 되는 것을 볼 수 있다.

# End(2022년 10월 18일)